src/network/networkd-routing-policy-rule.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+/***
+  This file is part of systemd.
+
+  Copyright 2017 Susant Sahani
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <inttypes.h>
+#include <stdbool.h>
+
+#include "in-addr-util.h"
+
+typedef struct RoutingPolicyRule RoutingPolicyRule;
+
+#include "networkd-link.h"
+#include "networkd-network.h"
+
+typedef struct Network Network;
+typedef struct Link Link;
+typedef struct NetworkConfigSection NetworkConfigSection;
+
+struct RoutingPolicyRule {
+        Network *network;
+        Link *link;
+        NetworkConfigSection *section;
+
+        uint8_t tos;
+
+        uint32_t table;
+        uint32_t fwmark;
+        uint32_t fwmask;
+        uint32_t priority;
+
+        int family;
+        unsigned char prefixlen;
```

Comment:
Agreed, that just flushing rules on shutdown is problematic. But I don't think it's possible to be "authoritative", because the rule table is global, and we cannot automatically determine which rules have been added by us just by looking at the rules. And we cannot flush all rules, because that would break routing completely.

One option, which is ugly, but might work, is to store the list of rules being written by us to a file in /run/systemd, and when starting up, to flush all rules listed there. It might be enough to just store the rule numbers for rules that we add ourselves (if the netlink reply message contains the rule number, I don't know).